### PR TITLE
[cxx-interop] Add SIL function representation cxx_method; Support extending C++ types.

### PR DESCRIPTION
**Pull request by zoecarver**
There are three major changes here:
 1. The addition of `SILFunctionTypeRepresentation::CXXMethod`.
 2. C++ methods are imported with their members *last*. Then the arguments are switched when emitting the IR for an application of the function.
 3. Clang decls are now marked as foreign witnesses.

These are all steps towards being able to have C++ protocol conformance.

A few other notes:
* There is a surprising lack of C++ method tests. As a separate PR, I'll add some shortly. 
* This will also let me write some tests for the witness layout optimizations (as [requested here](https://github.com/apple/swift/pull/32291#discussion_r532613678)).
* Protocol conformance is also needed for some future generics work.
* I'm hoping to submit future patches to add support for static C++ methods and operators. 

Resolves [SR-12750](https://bugs.swift.org/browse/SR-12750).

### CODE REVIEW

File: include/swift/AST/ExtInfo.h [hunk 1]
```diff
@@ -167,6 +167,9 @@ enum class SILFunctionTypeRepresentation : uint8_t {
 
   /// A closure invocation function that has not been bound to a context.
   Closure,
+
+  /// A C++ method that takes a "self" argument (not a static C++ method or cosntructor).
```
Comment (varungandhi-apple) [inline include/swift/AST/ExtInfo.h:171]: Typo: "cosntructor" -> constructor.

File: lib/AST/ASTMangler.cpp [hunk 2]
```diff
@@ -1693,6 +1693,7 @@ void ASTMangler::appendImplFunctionType(SILFunctionType *fn) {
       OpArgs.push_back('B');
       appendClangTypeToVec(OpArgs);
       break;
+    case SILFunctionTypeRepresentation::CXXMethod:
```
Comment (varungandhi-apple) [inline lib/AST/ASTMangler.cpp:1696]: Please update `docs/ABI/Mangling.rst` for this addition.

File: lib/SIL/IR/SILFunctionType.cpp [hunk 3]
```diff
@@ -4233,6 +4245,7 @@ TypeConverter::getLoweredFormalTypes(SILDeclRef constant,
     bridgedResultType = resultType;
     break;
 
+  case SILFunctionTypeRepresentation::CXXMethod:
```
Comment (varungandhi-apple) [inline lib/SIL/IR/SILFunctionType.cpp:4248]: In the body of the case, you could simplify the code to check if the representation is `CXXMethod` instead of doing `auto method = dyn_cast_or_null<clang::CXXMethodDecl>(constant.getDecl()->getClangDecl())`.

File: lib/SIL/IR/Bridging.cpp [hunk 4]
```diff
@@ -187,6 +188,7 @@ Type TypeConverter::getLoweredCBridgedType(AbstractionPattern pattern,
     case SILFunctionType::Representation::Thin:
     case SILFunctionType::Representation::Method:
     case SILFunctionType::Representation::ObjCMethod:
+    case SILFunctionTypeRepresentation::CXXMethod:
```
Comment (varungandhi-apple) [inline lib/SIL/IR/Bridging.cpp:191]: Not part of this PR, but the comment above looks a tad outdated.

```
// Functions that are already represented as blocks or C function pointers
// don't need bridging.
```

Potentially want to update that (or remove it) since you are updating nearby stuff.

File: lib/IRGen/GenCall.cpp [hunk 5]
```diff
@@ -3246,14 +3268,18 @@ static void externalizeArguments(IRGenFunction &IGF, const Callee &callee,
                 == SILFunctionTypeRepresentation::Block) {
     // Ignore the physical block-object parameter.
     firstParam += 1;
-    // Or the indirect result parameter.
+  } else if (callee.getRepresentation() ==
+             SILFunctionTypeRepresentation::CXXMethod) {
+    // Skip the "self" param.
+    paramEnd--;
+    params = params.drop_front();
```
Comment (varungandhi-apple) [inline lib/IRGen/GenCall.cpp:3275]: This reads in a very confusing manner because dropping from the front makes it look like the start should be updated and not the end. Is this logically equivalent to the `firstParam += 1` handling in the block case? That doesn't modify `params`, so I'm not sure why we are modifying `params` here.

File: lib/SIL/IR/SILFunctionType.cpp (hunk 3 above)
Comment (zoecarver) [reply lib/SIL/IR/SILFunctionType.cpp:4248]: We need `method`, so we'd end up casting it anyway, so I'm not sure how much better it would be. But happy to update if you'd still like me to.

File: lib/SIL/IR/Bridging.cpp (hunk 4 above)
Comment (zoecarver) [reply lib/SIL/IR/Bridging.cpp:191]: Hmm. Is this comment saying "don't need to be bridged here" or "don't need bridging in general"? The latter is outdated but the former seems OK. Also happy to remove it.

File: lib/IRGen/GenCall.cpp (hunk 5 above)
Comment (zoecarver) [reply lib/IRGen/GenCall.cpp:3275]: Heh, this is why we need some more C++ method tests. The `params.drop_front()` part isn't supposed to be there. Remember, the self parameter is now last. We could drop the last param, but that doesn't seem necessary because we're going to have to update `paramEnd` anyway. 

Maybe `paramCount` would be a better name, though.

**Comment by zoecarver** [comment]
@varungandhi-apple all resolved. Thanks for the review!

File: docs/ABI/Mangling.rst [hunk 6]
```diff
@@ -531,8 +531,8 @@ Types
   FUNCTION-KIND ::= 'B'                      // objc block function type
   FUNCTION-KIND ::= 'zB' C-TYPE              // objc block type with non-canonical C type
   FUNCTION-KIND ::= 'L'                      // objc block function type with canonical C type (escaping) (DWARF only; otherwise use 'B' or 'zB' C-TYPE)
-  FUNCTION-KIND ::= 'C'                      // C function pointer type
-  FUNCTION-KIND ::= 'zC' C-TYPE              // C function pointer type with with non-canonical C type
+  FUNCTION-KIND ::= 'C'                      // C function pointer / method type
+  FUNCTION-KIND ::= 'zC' C-TYPE              // C function pointer / method type with with non-canonical C type
```
Comment (varungandhi-apple) [inline docs/ABI/Mangling.rst:535]: Please use "C++ method" (otherwise, it's not clear if this is referring to ObjC methods or C++ methods).

Comment (zoecarver) [reply docs/ABI/Mangling.rst:535]: Thanks, done.

**Comment by zoecarver** [comment]
@swift-ci please smoke test.

**Comment by zoecarver** [comment]
@swift-ci please smoke test.

**Comment by zoecarver** [comment]
@swift-ci please test Windows.

**Comment by varungandhi-apple** [comment]
Mostly LGTM, deferring to other folks for specifics.

**Comment by zoecarver** [comment]
@swift-ci please smoke test.

**Comment by zoecarver** [comment]
@swift-ci please smoke test.

**Comment by zoecarver** [comment]
@swift-ci please smoke test.

**Comment by zoecarver** [comment]
@rjmccall and @hlopko friendly ping.

File: lib/IRGen/GenCall.cpp [hunk 7]
```diff
@@ -2187,7 +2191,12 @@ class SyncCallEmission final : public CallEmission {
       break;
 
     case SILFunctionTypeRepresentation::Block:
-      adjusted.add(getCallee().getBlockObject());
+    case SILFunctionTypeRepresentation::CXXMethod:
+      if (getCallee().getRepresentation() ==
```
Comment (hyp) [inline lib/IRGen/GenCall.cpp:2195]: Why are you handling the block and cxxmethod in same case if you need to do a different call for block vs cxx method? I think it would be better to have separate cases.

File: lib/SILGen/SILGenType.cpp [hunk 8]
```diff
@@ -417,9 +417,18 @@ template<typename T> class SILGenWitnessTable : public SILWitnessVisitor<T> {
     // If it's not an accessor, just look for the witness.
     if (!reqAccessor) {
       if (auto witness = asDerived().getWitness(reqDecl)) {
-        return addMethodImplementation(
-            requirementRef, requirementRef.withDecl(witness.getDecl()),
-            witness);
+        auto newDecl = requirementRef.withDecl(witness.getDecl());
+        // Don't import Objective-C methods as forien. If the following
```
Comment (hyp) [inline lib/SILGen/SILGenType.cpp:421]: Typo: foreign.

File: lib/SILGen/SILGenType.cpp [hunk 9]
```diff
@@ -417,9 +417,18 @@ template<typename T> class SILGenWitnessTable : public SILWitnessVisitor<T> {
...
+        // Objective-C function is imported here as foreing:
+        //   () -> String
+        // It will be imported as the following type:
+        //   () -> NSString
+        // But the first is correct, so make sure we don't mark this witness
+        // as forien.
```
Comment (hyp) [inline lib/SILGen/SILGenType.cpp:427]: Typo: foreign.

File: lib/SILGen/SILGenType.cpp [hunk 10]
```diff
@@ -417,9 +417,18 @@ template<typename T> class SILGenWitnessTable : public SILWitnessVisitor<T> {
     // If it's not an accessor, just look for the witness.
     if (!reqAccessor) {
       if (auto witness = asDerived().getWitness(reqDecl)) {
-        return addMethodImplementation(
-            requirementRef, requirementRef.withDecl(witness.getDecl()),
-            witness);
+        auto newDecl = requirementRef.withDecl(witness.getDecl());
+        // Don't import Objective-C methods as forien. If the following
+        // Objective-C function is imported here as foreing:
```
Comment (hyp) [inline lib/SILGen/SILGenType.cpp:422]: Do you mean a C++ method here instead of Objective-C function?

File: lib/Serialization/Serialization.cpp [hunk 11]
```diff
@@ -3906,6 +3906,7 @@ static uint8_t getRawStableSILFunctionTypeRepresentation(
   SIMPLE_CASE(SILFunctionTypeRepresentation, CFunctionPointer)
   SIMPLE_CASE(SILFunctionTypeRepresentation, Method)
   SIMPLE_CASE(SILFunctionTypeRepresentation, ObjCMethod)
+  SIMPLE_CASE(SILFunctionTypeRepresentation, CXXMethod)
   SIMPLE_CASE(SILFunctionTypeRepresentation, WitnessMethod)
   SIMPLE_CASE(SILFunctionTypeRepresentation, Closure)
```
Comment (hyp) [inline lib/Serialization/Serialization.cpp:3911]: The new case should go after `Closure` to match the enum above, right?

**Comment by zoecarver** [comment]
@swift-ci please smoke test.

File: lib/IRGen/GenCall.cpp (hunk 7 above)
Comment (zoecarver) [reply lib/IRGen/GenCall.cpp]: Aha. Turns out there was actually a reason for this: we want the fall-through here for both cases. Without this weird if-inside-the-case pattern we don't `externalizeArguments` for `::Block`.

**Comment by zoecarver** [comment]
@swift-ci please smoke test.

**Comment by zoecarver** [comment]
@swift-ci please smoke test.

**Comment by zoecarver** [comment]
@swift-ci please smoke test.

File: lib/SILGen/SILGenType.cpp [hunk 12]
```diff
@@ -415,8 +415,19 @@ template<typename T> class SILGenWitnessTable : public SILWitnessVisitor<T> {
...
+        // Only import C++ methods as foreign. If the following
+        // Objective-C function is imported as foreign:
+        //   () -> String
+        // It will be imported as the following type:
+        //   () -> NSString
+        // But the first is correct, so make sure we don't mark this witness
+        // as foreign.
+        if (dyn_cast_or_null<clang::CXXMethodDecl>(
+                witness.getDecl()->getClangDecl()))
+          newDecl = newDecl.asForeign();
```
Comment (slavapestov) [inline lib/SILGen/SILGenType.cpp:428]: Why are C++ methods different than Objective-C methods in this regard?

Comment (zoecarver) [reply lib/SILGen/SILGenType.cpp:428]: Basically, if an Objective-C method is marked as foreign it will be imported using the Objective-C type and not the Swift type. For example, it might be imported as `() -> NSString` instead of `() -> String` (which is what we want).

**Review by hyp — APPROVED** [review]
LGTM
